integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.ts

```diff
@@ -0,0 +1,1496 @@
+  }
+
+  private _leafNodeRefs: NodeRef[];
+  private get leafNodeRefs(): NodeRef[] {
+    return this._leafNodeRefs;
+  }
+  private set leafNodeRefs(value: NodeRef[]) {
+    this._leafNodeRefs = value;
+  }
+
+  private _viewPriorToIsolateViewInfo: any;
+  private get viewPriorToIsolateViewInfo(): any {
+    return this._viewPriorToIsolateViewInfo;
+  }
+  private set viewPriorToIsolateViewInfo(value: any) {
+    this._viewPriorToIsolateViewInfo = value;
+  }
+
+  private _viewportAdded$: Observable<void>;
+  private get viewportAdded$(): Observable<void> {
+    return this._viewportAdded$;
+  }
+  private set viewportAdded$(value: Observable<void>) {
+    this._viewportAdded$ = value;
+  }
+
+  private _selectedNodeIds$ = new BehaviorSubject<string[]>([]);
+  private get selectedNodeIds$() {
+    return this._selectedNodeIds$;
+  }
+  private set selectedNodeIds$(value) {
+    this._selectedNodeIds$ = value;
+  }
+
+  private _sceneLoadInfo$ = new BehaviorSubject<SceneLoadInfo>({
+    sceneLoadState: SceneLoadState.NotStarted,
+  });
+  public get sceneLoadInfo$() {
+    return this._sceneLoadInfo$;
+  }
+
+  protected readonly DEFAULT_BACKGROUND_TOP_COLOR = '--cx-color-inverse';
+  protected readonly DEFAULT_BACKGROUND_BOTTOM_COLOR = '--cx-color-inverse';
+  protected readonly DEFAULT_HOTSPOT_SELECTION_HIGHLIGHT_COLOR =
+    'rgba(255, 0, 0, 0.6)';
+  protected readonly DEFAULT_SHOW_ALL_HOTSPOTS_COLOR = 'rgba(255, 255, 0, 0.3)';
+  protected readonly DEFAULT_OUTLINE_COLOR = 'red';
+  protected readonly DEFAULT_OUTLINE_WIDTH = 5;
+  protected readonly DEFAULT_SELECTION_MODE = SelectionMode.Exclusive;
+  protected readonly DEFAULT_SHOW_ALL_HOTSPOTS_ENABLED = false;
+  protected readonly DEFAULT_EXCLUDED_OPACITY = 0.2;
+  protected readonly DEFAULT_ZOOM_TO_MARGIN = 0.2;
+  protected readonly DEFAULT_FLY_TO_DURATION = 1;
+
+  private _flyToDurationInSeconds = this.DEFAULT_FLY_TO_DURATION;
+  private get flyToDurationInSeconds() {
+    return this._flyToDurationInSeconds;
+  }
+  private set flyToDurationInSeconds(value) {
+    this._flyToDurationInSeconds = value;
+  }
+  private _zoomToMargin = this.DEFAULT_ZOOM_TO_MARGIN;
+  private get zoomToMargin() {
+    return this._zoomToMargin;
+  }
+  private set zoomToMargin(value) {
+    this._zoomToMargin = value;
+  }
+
+  /**
+   * The top colour of the background gradient.
+   * Can be passed in the CSS color format or as a Spartacus theme color i.e. '--cx-color-background' with the quotes.
+   */
+  @Input()
+  public set backgroundTopColor(backgroundTopColor: string) {
+    if (this._backgroundTopColor === backgroundTopColor) {
+      return;
+    }
+    this._backgroundTopColor = backgroundTopColor;
+    this.executeWhenSceneLoaded(() => {
+      this.viewport.setBackgroundColorTop(this.getCSSColor(backgroundTopColor));
+    });
+  }
+  public get backgroundTopColor(): string {
+    return this._backgroundTopColor;
+  }
+  private _backgroundTopColor: string;
```

Comment:
Extensibility is important for customers. They are extending and customizing our services' classes. For them it's important to access our methods and properties in their child classes. 

What do you think about changing all private properties in this service to protected ones? 
Or maybe you deliberately wanted to hide some logic as an non-customizable implementation detail? (e.g. because it's a potential subject of a breaking change in the future)?

Similar comment applies also for other services in this PR.

Reply:
There is a protected field that allows a default value to be specified here:
https://github.com/SAP/spartacus/blob/a3be8ca871c26bf8aba0a4dbfc59b81a6270f79c/integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.ts#L220

I think it is useful to limit access to this field so that it is only modified via the getters and setters of the corresponding public property (which relates to a value that can only be set once a scene has loaded).